src/utils/generic.js

```diff
@@ -0,0 +1,308 @@
+/* eslint-disable max-classes-per-file */
+/**
+ * A `Map` that has a capacity.
+ */
+export class MapWithCapacity extends Map {
+  /**
+   *
+   * @param {number} capacity
+   */
+  constructor(capacity, ...args) {
+    if (capacity < 1) {
+      throw new Error(`Capacity must be greater than 0, received: ${capacity}`);
+    }
+    super(...args);
+    this.capacity = capacity;
+    this.adjustSize();
+  }
+
+  set(key, value) {
+    this.delete(key);
```

Comment:
This is needed to make the correct order of the ' entries`/`values`/`keys` iterators that produce items in the insertion order.
If a key already exists in the map, it won't change its position in case of value override, however, we need to have the last inserted entry being at the end of the iteration.

Reply:
Document this would be helpful.